### DISCUSSION TITLE
Get provider access token ONLY server side

### DISCUSSION THREAD

**Discussion by acerbisgianluca** [nextauthjs/next-auth#5490] [Help] [answered]
Hey there! I'm using a custom Oauth provider and Dynamodb adapter.

I would like to access the provider access token only server-side, without exposing it to the client.

I read [this tutorial](https://next-auth.js.org/tutorials/refresh-token-rotation) about refresh token rotation, but:

1- `jwt` callback isn't called since I'm using a DB (session strategy is database)
2- `session` callback adds the access token to the session, so it's available client-side. I think this is a bad security practice, isn't it?

At this point, the only solution that comes to my mind is to manually query the database for the access token (the adapter doesn't provide a method for this either). Do you have any other idea?

**Comment by TomFreudenberg** [comment]
Hi,

why don't you use "jwt" as strategy? You may set that explicit.

> ccess token to the session, so it's available client-side

something must be client available to know about the session?

using "jwt" it will be in a secure session cookie.

hopefully I am right with that information.

**Reply by acerbisgianluca** [reply]
Hi Tom, thank you for the answer.

With the jwt strategy, the provider access token would be added to the jwt payload, but that payload is publicly viewable from everyone who can access the cookie storage, right?

Just for a bit of context, here is what I had in mind:
- client calls my Next.js API (`pages/api/...`)
- the API checks if the user is logged in and gets the access token from the provider
- the API calls the provider's API with the access token

So for the second step I need a way for the server (not the client) to get the provider's access token from the database. I wouldn't like to call the provider's API directly from the client code.

**Reply by TomFreudenberg** [reply, marked as answer]
Ok, got that.

In the case that you won't expose the providers API token - you are right - and you need to store it somewhere.

JWT may be the option - you may be allowed to store anything in it. So what about hashing (crypto by a server salt) the providers token and put it in the JWT. Nobody with knowledge of your salt can export it.

Maybe easier than writing some database queries and management ...
